rp2: Add support for new RP2350 MCU.

dpgeorge opened the pull request:
### Summary

This PR adds support for the new RP2350 MCU, and the new Pico 2 board.

Both the Cortex-M33 and RISC-V CPUs are supported.  You can build for them both (as separate firmware) with:
```
make BOARD=RPI_PICO2
make BOARD=RPI_PICO2 BOARD_VARIANT=RISCV
```

Download the .uf2 to the board as usual via mass storage.  The bootloader will detect which architecture (ARM or RISC-V) the firmware has and boot the correct CPU(s).

See https://www.raspberrypi.com/products/rp2350/

Thanks to @peterharperuk and @Gadgetoid for doing a lot of the work in this PR.

### Testing

The existing test suite (mostly) passes on the Pico 2, in both ARM and RISC-V mode.  There are some issues with lightsleep mode on ARM and machine-idle on both ARM and RISC-V which need to be sorted out.

### TODO

- [ ] ~~fix lightsleep~~ will do at a later date when there's a new version of pico-sdk with required improvements
- [x] get all tests passing with RPI_PICO2 and RPI_PICO2-RISCV firmware
- [x] work out why RPI_PICO_W build has increased so much -> because of time zone support in libc, fixed by providing custom `localtime_r` and `mktime`
- [x] test RPI_PICO and RPI_PICO_W
- [x] ~~test other TinyUSB ports~~ no longer needed, TinyUSB already updated in #15902
- [x] decide on board name: `RPI_PICO2` vs `RPI_PICO_2` (latter matches `RPI_PICO_W`, and officially it's `Pico 2`) -> go with `RPI_PICO2` to match pico-sdk header `pico2.h`

peterharperuk committed:
rp2/boards/RPI_PICO2: Add new RPI_PICO2 board definition.

This is the same form-factor as the Pico but with an RP2350.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
rp2/modmachine: Implement lightsleep for RP2350.

This isn't fully working, the CPU often wakes up early.  That will be fixed
when a newer version of pico-sdk is released.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
rp2/mpconfigport: Set MCU name for RP2350.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
rp2: Update custom linker scripts for new pico-sdk.

Signed-off-by: Phil Howard <phil@gadgetoid.com>
Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
rp2: Integrate RP2350 and use aon_timer instead of rtc API.

This commit separates various build settings and include files that are
specific to RP2040 and RP2350, and uses the aon_timer interface instead of
rtc, to work across both MCU variants.

Signed-off-by: Damien George <damien@micropython.org>
Signed-off-by: Phil Howard <phil@gadgetoid.com>

peterharperuk committed:
rp2/rp2_dma: Generalise DMA for RP2350.

Two new bits were added to the ctrl register, and existing bits were
shifted, so use DMA_CH0_CTRL_TRIG_xxx constants to generalise the code.

Signed-off-by: Damien George <damien@micropython.org>

Gadgetoid committed:
py/usermod.cmake: Check target exists in usermod_gather_sources.

Check a target exists before accessing properties.  Otherwise
usermod_gather_sources would recurse into garbage property names and break.

Signed-off-by: Phil Howard <phil@gadgetoid.com>

Gadgetoid committed:
rp2: Add support for 48-pin RP2350 variant.

Update NUM_GPIOS to match NUM_BANK0_GPIOS, and increase bit-width of
variables that store pin numbers.

Signed-off-by: Phil Howard <phil@gadgetoid.com>

Gadgetoid committed:
rp2/rp2_pio: Add support for RP2350A/B variants in PIO interface.

Add support for 32 and 48 pin variants of RP2350.

Add new `PIO.gpio_base()` method, mirroring the Pico SDK.

Signed-off-by: Phil Howard <phil@gadgetoid.com>
Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
rp2/boards/make-pins.py: Pass num-gpios/num-ext-gpios into make-pins.

NUM_GPIOS amd NUM_EXT_GPIOS are currently hardcoded in make-pins.py, which
makes it difficult to support SoCs with different pin count.

This commit generalises make-pins.py by passing in the pin count in via the
new arguments `--num-gpios` and `--num-ext-gpios`.  These default to the
current values supported by Pico, namely 30/10.  This can be changed with
PICO_NUM_GPIOS and PICO_NUM_EXT_GPIOS in `mpconfigboard.cmake`.

Signed-off-by: Damien George <damien@micropython.org>

Gadgetoid committed:
rp2/machine_pin: Generalise gpio_irq handler for pins >32.

Fix the gpio_irq function so that it looks at all six iobank0_hw->intr[n]
registers, for up to 48 IOs.

Signed-off-by: Phil Howard <phil@gadgetoid.com>

peterharperuk committed:
rp2/machine_pin: Use 64-bit gpio functions to allow gpios >=32 to work.

Signed-off-by: Damien George <damien@micropython.org>

peterharperuk committed:
rp2/clocks_extra: Update runtime_clocks_init based on new pico-sdk.

Signed-off-by: Damien George <damien@micropython.org>

sfe-SparkFro committed:
rp2/machine_bitstream: Set SysTick reset value.

In case it doesn't have the correct value.

Signed-off-by: Damien George <damien@micropython.org>

Gadgetoid committed:
rp2/machine_adc: Add ADC support for RP2350B.

Signed-off-by: Damien George <damien@micropython.org>

sfe-SparkFro committed:
rp2/machine_uart: Allow new TX/RX pins on RP2350.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/ports/rp2: Update DMA test to work on RP2350.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
lib/pico-sdk: Update to version 2.0.0.

Adds support for the new RP2350 MCU.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
rp2: Add support for RP2350 in RISCV mode.

As part of this change, the RV32I native emitter is enabled on RISCV
board variants.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
rp2/Makefile: Allow CMAKE_ARGS to be set by user.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
rp2/mphalport: Implement mp_hal_ticks_cpu for RISCV using mcycle.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
rp2/machine_bitstream: Implement bitstream for RISC-V using mcycle.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Pre-built firmware can be found here: https://micropython.org/download/RPI_PICO2/

dpgeorge committed:
tests/ports/rp2: Add simple rp2-specific UART test.

To test construction of UART instances.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tools/ci.sh: Add RPI_PICO2 to CI.

Signed-off-by: Damien George <damien@micropython.org>

Gadgetoid commented:
> work out why RPI_PICO_W build has increased so much

Build size, or build time? Do we need to set `-DPICO_NO_COPRO_DIS=1`?

dpgeorge commented:
> Build size, or build time? Do we need to set `-DPICO_NO_COPRO_DIS=1`?

Build/firmware size.

Setting `DPICO_NO_COPRO_DIS=1` does not seem to change anything (on RP2040 builds at least).

dpgeorge commented:
> work out why RPI_PICO_W build has increased so much

It looks like this is due to the new always-on timer (aon timer) code, which calls `localtime_r()`, which subsequently pulls in timezone support and calls to large printf and scanf functions (from the C stdlib, newlib).

It doesn't look easy to remove all that code even though it's probably never used by MicroPython.  So we may have to live with it.

Gadgetoid commented:
Hmm so `aon_timer_set_time()` calls `time_to_datetime()` from `pico_util`.

It might be possible to provide our own `pico_util` with a more MicroPython-friendly alternative for that function, or just to shim `localtime_r`, eg:

```cmake
target_link_options(${MICROPY_TARGET} PRIVATE "LINKER:--wrap=localtime_r")
```

```c++
bool __wrap_localtime_r(time_t time, struct tm local_time) {
    gmtime_r(&time, &local_time);
    return false;
}
```

Which I think resolves to `mbedtls_platform_gmtime_r`

kilograham commented:
Ah, yes, this might be a vote against using the aon_timer API on RP2040 after all - the POWMAN timer is not date based, but the RTC one is, so one of them takes the hit for a common API :-( ... we should probably add both forms for both though (and maybe a documentation warning)

projectgus committed:
rp2: Workaround pico_aon_timer timezone binary size increase.

Provide stub implementations of localtime_r() and mktime() to avoid
code size increase.

Reported upstream at https://github.com/raspberrypi/pico-sdk/issues/1810

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Here's an extended version of @Gadgetoid's patch that gets the code size all the way down (PICO_W binary size is actually slightly smaller than the current master). I have verified this compiles and boots but I have not tested it further:
https://github.com/projectgus/micropython/commit/debc0ad6c1a528d9ecf97735a3a0e2bcfdefa032

(Two tips for anyone who is interested: 1. Linker wrap is actually not necessary for any symbol that is linked from a library, as any object provided on the linker command line will take precedence. [Great blog post on the details](https://eli.thegreenplace.net/2013/07/09/library-order-in-static-linking). 2. The --cref linker option is worth its weight in gold for figuring this stuff out. I will submit a PR to add it either to pico-sdk or rp2 port linker command lines.)

projectgus reviewed (commented):
This is an impressively small and clean set of changes for adding not just a new chip but also a new CPU arch! Kudos to the pico-sdk team and everyone who worked on this PR.

Gadgetoid commented on ports/rp2/rp2_pio.c:
```diff
@@ -357,6 +381,30 @@ static mp_obj_t rp2_pio_state_machine(size_t n_args, const mp_obj_t *pos_args, m
+
+    uint8_t gpio_base = mp_obj_get_int(gpio_base_in);
+
+    // TODO check for RP2350B?
```
Aha, `gpio_base` is only applicable on RP2350B where there are 48 IOs instead of 32. It can be 0 or 16 to give the PIO access to pins 0-31 or 16-47 respectively.

As such this function should do something reasonable, like raise a ValueError on RP2040 or RP235XA hardware.

We should probably also check the value is one of the two valid inputs, but it started to feel like a lot of input validation for a currently slightly obscure (albeit Pimoroni have shipped a Pico-a-like with a B variant chip) method.

dpgeorge replied:
I've now removed this todo, and made this method available only on RP2350B.  It's a pretty special method and must be called before any PIO programs are added (or StateMachine's created).  I've also added docs for this new `PIO.gpio_base()` method to describe its use.

Gadgetoid commented:
> Here's an extended version of @Gadgetoid's patch that gets the code size all the way down

I've said it once, I'll say it again, Open Source is just the formal version of posting the wrong answer on the internet to get the right one :laughing:  thank you!

I've just run our batteries-included builds against this branch and that extra 17k blows our flash budget :grimacing:

ZodiusInfuser commented:
Hey. I just wanted to chime in here to point out that one of our PGA2350 users (which uses the B chip) was not able to use PWM on the pins above 31. Checking this repo, it seems `machine_pwm` has no additions for the extra 8 channels yet.

Gadgetoid committed:
rp2/machine_pwm: Add RP2350 slices to machine.PWM.

Signed-off-by: Phil Howard <phil@gadgetoid.com>

Gadgetoid commented:
Afacit PWM just needs this: https://github.com/pimoroni/micropython/commit/a732af7b15d81dfa3d87203b6cc19701b1c9e48f

I'll cut some builds to test it!

ZodiusInfuser commented:
Thanks @Gadgetoid . Looks like there's at least one hardcoded slice count left:
```python
void machine_pwm_deinit_all(void) {
    for (int i = 0; i < 8; i++) {
        slice_freq_set[i] = false;
        pwm_set_enabled(machine_pwm_obj[i].slice, false);
    }
}
```

Gadgetoid commented:
Good spot! Curse the magic numbers! Pushed a fix.

mattytrentini commented:
Built `RPI_PICO2` and deployed it to the [Seeed XIAO RP2350](https://www.seeedstudio.com/Seeed-XIAO-RP2350-p-5944.html); it boots and at least some basic operations appear to be working fine! 😄 

```py
> mpremote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell
MicroPython v1.24.0-preview.203.gb9f6698d30.dirty on 2024-08-23; Raspberry Pi Pico2 with RP2350
Type "help()" for more information.
>>> from machine import Pin
>>> # Check user LED
>>> led = Pin(25, Pin.OUT)
>>> led.toggle()
>>> # Check RGB LED
>>> power_pin = Pin(23, Pin.OUT) # Provide power first
>>> power_pin.value(1)
>>> from neopixel import NeoPixel
>>> neo = NeoPixel(Pin(22), 1)
>>> neo[0] = (0, 0, 40)
>>> neo.write() # Blue LED
```

dpgeorge commented on shared/tinyusb/mp_usbd_runtime.c:
```diff
@@ -429,6 +429,8 @@ void mp_usbd_init(void) {
     if (need_usb) {
         tusb_init(); // Safe to call redundantly
+        tud_cdc_configure_fifo_t cfg = { .rx_persistent = 0, .tx_persistent = 1 };
+        tud_cdc_configure_fifo(&cfg);
```
Done, added `mp_usbd_init_tud()` helper.

dpgeorge commented on tests/ports/rp2/rp2_uart.py:
```diff
@@ -0,0 +1,10 @@
+
+print(UART(0, tx=0, rx=1))
+
+if "RP2350" in "sys.implementation._machine":
```
Oops!  Fixed.

dpgeorge commented:
> Here's an extended version of @Gadgetoid's patch that gets the code size all the way down (PICO_W binary size is actually slightly smaller than the current master).

Thanks for that, I've now added that commit to this PR.

dpgeorge commented:
> decide on board name: `RPI_PICO2` vs `RPI_PICO_2` (latter matches `RPI_PICO_W`, and officially it's `Pico 2`)

The decision is to stick with `RPI_PICO2`, and then also `RPI_PICO2_W` (eventually), because that matches the pico-sdk header file names: `pico.h`, `pico_w.h`, `pico2.h` and `pico2_w.h`

Gadgetoid commented:
Looks like I have some PSRAM rebasing to do! Will sure make it easier to play with the new UART/PPP stuff 😁

(I know you’re probably still working your way down here, but don’t miss the PWM patch above! - https://github.com/pimoroni/micropython/commit/7f1fcc72dbe9b39dda05b15d9e60a04eb2aa5068 )

Gadgetoid commented:
Might be my board config, but I run into this when trying to build for RP2350B or `PICO_NUM_GPIOS=48`-

```
/home/runner/work/pimoroni-pico-rp2350/pimoroni-pico-rp2350/micropython/ports/rp2/rp2_pio.c: In function 'rp2_pio_gpio_base':
/home/runner/work/pimoroni-pico-rp2350/pimoroni-pico-rp2350/micropython/ports/rp2/rp2_pio.c:411:48: error: 'pin_GPIO0' undeclared (first use in this function)
  411 |     return pio_get_gpio_base(self->pio) == 0 ? pin_GPIO0 : pin_GPIO16;
      |                                                ^~~~~~~~~
```

dpgeorge committed:
rp2/machine_pin: Move decl of machine_pin_obj_table to public header.

So other code can include `machine_pin.h` and use the pin name macros such
as `pin_GPIO0`.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> I know you’re probably still working your way down here, but don’t miss the PWM patch above!

Thanks for that, I checked your patch and cherry-picked it to this PR.

> Might be my board config, but I run into this when trying to build for RP2350B or `PICO_NUM_GPIOS=48`

Aah, sorry about that build error, now fixed and tested on a Pico Plus2 (and thanks for that hardware!).

MichaelBell commented:
You might want to add PICO_EMBED_XIP_SETUP=1 to the compile definitions: https://github.com/MichaelBell/micropython/commit/87208bc2053e2c70cd84bc7087f6b451e777a3b2

Without this the flash setup (what would have been the second stage bootloader on RP2040) isn't run.  So the flash timing is just the default from the bootrom, and it isn't put into continuous read mode.

I see the following performance difference with this change:
```
diff of microsecond times (lower is better)
N=150 M=440                before.txt ->  after.txt         diff      diff% (error%)
bm_chaos.py                 167229.62 ->  150712.75 :  -16516.87 =  -9.877% (+/-0.06%)
bm_fannkuch.py               76832.00 ->   78101.38 :   +1269.38 =  +1.652% (+/-0.04%)
bm_fft.py                   288179.25 ->  302195.50 :  +14016.25 =  +4.864% (+/-0.04%)
bm_float.py                  66538.25 ->   51853.88 :  -14684.37 = -22.069% (+/-0.14%)
bm_hexiom.py                246255.50 ->  232976.25 :  -13279.25 =  -5.392% (+/-0.07%)
bm_nqueens.py               306577.75 ->  263226.62 :  -43351.13 = -14.140% (+/-0.07%)
bm_pidigits.py               83102.88 ->   82278.00 :    -824.88 =  -0.993% (+/-0.04%)
bm_wordcount.py             161100.88 ->  140273.50 :  -20827.38 = -12.928% (+/-0.04%)
core_import_mpy_multi.py    131755.25 ->  106147.88 :  -25607.37 = -19.436% (+/-0.14%)
core_import_mpy_single.py    12673.12 ->   11258.50 :   -1414.62 = -11.162% (+/-0.36%)
core_locals.py              185487.88 ->  175717.12 :   -9770.76 =  -5.268% (+/-0.06%)
core_qstr.py                 23146.38 ->   21837.00 :   -1309.38 =  -5.657% (+/-0.05%)
core_str.py                 241620.12 ->  235245.88 :   -6374.24 =  -2.638% (+/-0.05%)
core_yield_from.py           63448.75 ->   63265.38 :    -183.37 =  -0.289% (+/-0.02%)
misc_aes.py                  68263.38 ->   70737.50 :   +2474.12 =  +3.624% (+/-0.06%)
misc_mandel.py              132251.88 ->  131381.88 :    -870.00 =  -0.658% (+/-0.04%)
misc_pystone.py             161164.62 ->  148147.00 :  -13017.62 =  -8.077% (+/-0.08%)
misc_raytrace.py            157036.88 ->  142396.38 :  -14640.50 =  -9.323% (+/-0.05%)
viper_call0.py               63094.75 ->   62816.50 :    -278.25 =  -0.441% (+/-0.01%)
viper_call1a.py              64676.75 ->   64399.88 :    -276.87 =  -0.428% (+/-0.00%)
viper_call1b.py              79088.12 ->   78809.50 :    -278.62 =  -0.352% (+/-0.00%)
viper_call1c.py              77891.25 ->   77607.75 :    -283.50 =  -0.364% (+/-0.00%)
viper_call2a.py              65878.50 ->   65600.50 :    -278.00 =  -0.422% (+/-0.01%)
viper_call2b.py              89171.75 ->   88889.12 :    -282.63 =  -0.317% (+/-0.00%)
```

So generally a decent improvement - though I must admit I have no idea why some tests get slower!

dpgeorge commented:
> You might want to add PICO_EMBED_XIP_SETUP=1 to the compile definitions

Good idea, thanks!

> though I must admit I have no idea why some tests get slower!

On an MCU that relies a lot on (instruction) caching, this is inevitable.  Overall the performance is improved by a lot more than it goes down on those three tests.

dpgeorge commented:
I verified that enabling `PICO_EMBED_XIP_SETUP=1` also improves performance of the RP2350 in RISCV mode, and it improves more than the ARM case.

Also, enabling this option on ARM doesn't increase power consumption, but it does increase by about 1mA on RISCV (at an idle REPL).  I guess that's acceptable for the improved performance.

peterharperuk commented:
@dpgeorge Sorry I've been a bit lax with this. But I've been looking at why sleep isn't working on rp2350 (I'm a bit stumped to understand this got past me). I can't really explain why it appears to be working on rp2040 but on rp2350 it seems the device is waking up immediately due to a soft timer alarm. A minimal fix to mp_machine_lightsleep is to calculate the absolute alarm expiry time, use this for the alarm and then do the following...

```
absolute_time_t t = make_timeout_time_ms(delay_ms);
...
timer_hw->alarm[alarm_num] = t
...
while(absolute_time_diff_us(get_absolute_time(), t) > 0) __wfi();
```

I've tested this and I see power usage going down. I think this code could do with a bit more tidying. I don't much like that we hardcode the alarm to 3, as this is used for the alarm pool (which ideally we'd like you to go back to using). We could also consider dormat sleep or powman power down.

MichaelBell commented:
Since https://github.com/micropython/micropython/commit/d420b4e478267cf8c6c0136982bdeeedaccf3920 was merged, RP2350 now defaults to 125MHz not 150MHz.

Maybe it would be better to do `set_sys_clock_khz(mp_hal_get_cpu_freq()/1000, false);`

robert-hh commented:
The intention was to have a fixed clock state after hard reset and soft reset. If just the current clock value is used, then it will not be reset at soft reset. Maybe instead different default clock values should be used for RP2040 and RP2350.

MichaelBell commented:
Ah yes, it has always been a bit odd that soft reset doesn't reset the clock frequency.

I think this would be better as a define (defaulted appropriately for RP2040 and RP2350), then it could be overridden in the `mpconfigboard.h`, e.g. if you wanted to underclock for a low power board, or set a default overclock for something like PicoSystem.  Another define that prevented this from happening would also be nice, so that a build can be made with a custom clock setup without having to patch Micropython.

That's all fairly independent from RP2350.  I'm happy to make a small PR for it, and then once that is in this PR can be adjusted to add in the alternative default.

robert-hh commented:
Looking into the code, this call is outside of the soft reset loop. So using constants would not change the soft reset behavior. Your initially suggested change would work for hard reset without introducing new build constants. A main.py/boot.py script can always change the frequency if needed.

dpgeorge committed:
rp2/main: Set CPU frequency to default for the MCU.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I've now made the rp2350 builds default to 150MHz, using the new `SYS_CLK_KHZ` macro provided by the pico-sdk.

I also updated TinyUSB to the latest release 0.17.0.

> I'm happy to make a small PR for it, and then once that is in this PR can be adjusted to add in the alternative default.

It would be good to be able to configure the default CPU frequency from mpconfigboard.h.  Feel free to make a PR for that, but note the change above to use `SYS_CLK_KHZ` as the default (maybe a board can redefine that??).

dpgeorge commented:
> I also updated TinyUSB to the latest release 0.17.0.

Let's do that separately in #15902.

dpgeorge commented:
> But I've been looking at why sleep isn't working on rp2350. ... I can't really explain why it appears to be working on rp2040 but on rp2350 it seems the device is waking up immediately due to a soft timer alarm. A minimal fix to mp_machine_lightsleep is to calculate the absolute alarm expiry time, use this for the alarm and then do the following...
> 
> ```
> absolute_time_t t = make_timeout_time_ms(delay_ms);
> ...
> timer_hw->alarm[alarm_num] = t
> ...
> while(absolute_time_diff_us(get_absolute_time(), t) > 0) __wfi();
> ```

Thanks @peterharperuk for looking into the lightsleep issue.

Unfortunately I don't think the above fix (doing a loop over `__wfi()`) is a good fix, for two reasons:
- Interrupts are disabled here, so if an interrupt does come in then the wfi wakes up but the interrupt is not processed.  So when the wfi is executed again it will wake immediately.  Hence it will turn into a busy loop and not save power.
- `machine.lightsleep()` may and should return early if there are events that need processing.  Eg you may be waiting for a pin IRQ, in which case the `lightsleep` should return as soon as the pin IRQ is raised, even if the timeout is not yet reached.  For this to work it needs a single call to `__wfi()`.

dpgeorge commented:
> we could exit the loop if there's any interrupt pending while ignoring timer interrupts - except maybe for the one we're interested in.

That may fix the second issue (return early when needed) but won't fix the first, it'll still be a busy loop.

Do you know what interrupt is waking the wfi early?

peterharperuk commented:
It's a while a go I looked at this but I remember that it was the soft timer alarm interrupt. I'll go back and see if I can work out what's going on.

sk8board commented:
Are you aware the RP2350 has two new state machine instructions? 

These new instructions are found in the [Pico C SDK datasheet](https://datasheets.raspberrypi.com/pico/raspberry-pi-pico-c-sdk.pdf):



![image](https://github.com/user-attachments/assets/adbf5854-f485-4296-9514-6857e4a3b8cf)

![image](https://github.com/user-attachments/assets/c7222bca-a41b-425b-b1a5-ac54a75d2f17)

These are nice features. Are you planning to add these two new instructions to MicroPython?

dpgeorge committed:
tests/ports/rp2: Update lightsleep/machine_idle to skip on RP2350.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/run-tests.py: Only run inlineasm tests on rp2 ARM targets.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
rp2/machine_uart: Clear timeout_char when UART is first constructed.

Otherwise a previous value of `timeout_char` may be left over after a soft
reset.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> Are you aware the RP2350 has two new state machine instructions?

Yes.  We don't have the capacity to add these now, so will add them at a later date.

dpgeorge commented:
I've now updated the rp2 specific tests so they pass on RP2040, RP2350 and RP2350-RISCV boards.

Retesting Pico and Pico W boards with this PR, they still pass the test suite.

I think this PR is pretty much good to merge now!

dpgeorge commented:
This is finally merged!  Thanks to everyone here who contributed/helped/tested/reviewed.
